In [1]:
import pandas as pd

Для этого урока используется данные о качестве воздуха $NO_2$, данные предоставляются организацией [`openaq`](https://openaq.org/) и загружается с помощью модуля [`py-openaq`](http://dhhagan.github.io/py-openaq/index.html).

<img src="https://newdaynews.ru/pict/arts1/66/19/661966_b.jpg" height="500px" width="600px" >

Набор данных `air_quality_no2_long.csv` содержит значения $NO_2$ для измерительных станций `FR04014`, `BETR801` и `London Westminster` соответственно в Париже, Антверпене и Лондоне.

In [2]:
air_quality_no2 = pd.read_csv("data/air_quality_no2_long.csv",
                                parse_dates=True)

In [3]:
air_quality_no2 = air_quality_no2[["date.utc", "location",
                                    "parameter", "value"]]

In [4]:
air_quality_no2.head()

date.utc location parameter  value
0  2019-06-21 00:00:00+00:00  FR04014       no2   20.0
1  2019-06-20 23:00:00+00:00  FR04014       no2   21.8
2  2019-06-20 22:00:00+00:00  FR04014       no2   26.5
3  2019-06-20 21:00:00+00:00  FR04014       no2   24.9
4  2019-06-20 20:00:00+00:00  FR04014       no2   21.4

Для этого урока также используются данные о качестве воздуха для твердых частиц размером менее 2,5 микрометров, данные предоставляются организацией [`openaq`](https://openaq.org/) и загружается с помощью модуля [`py-openaq`](http://dhhagan.github.io/py-openaq/index.html).

см. [`Частицы РМ2.5: что это, откуда и почему об этом все говорят`](https://habr.com/ru/company/tion/blog/396111/)

<img src="https://habrastorage.org/files/9b1/c66/b89/9b1c66b89f85464b8365b77c9ecbe781.jpg" height="500px" width="600px" >

Набор данных `air_quality_pm25_long.csv` содержит значения $PM_25$ для измерительных станций `FR04014`, `BETR801` и `London Westminster` соответственно в Париже, Антверпене и Лондоне.

In [5]:
air_quality_pm25 = pd.read_csv("data/air_quality_pm25_long.csv",
                                parse_dates=True)

In [6]:
air_quality_pm25 = air_quality_pm25[["date.utc", "location",
                                     "parameter", "value"]]

In [7]:
air_quality_pm25.head()

date.utc location parameter  value
0  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0
1  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5
2  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5
3  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0
4  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5

### Как объединить данные из нескольких таблиц? 

<img src="https://pandas.pydata.org/pandas-docs/stable/_images/08_concat_row1.svg" >

Я хочу объединить измерения $NO_2$ и $PM_25$ с похожей структурой в одну таблицу:

In [8]:
air_quality = pd.concat([air_quality_pm25, air_quality_no2], axis=0)

In [9]:
air_quality.head()

date.utc location parameter  value
0  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0
1  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5
2  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5
3  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0
4  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5

Функция [`concat()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html#pandas.concat) выполняет операцию конкатенации нескольких таблиц вдоль одной оси (по строкам или столбцам).

По умолчанию конкатенация происходит вдоль `оси 0`, поэтому результирующая таблица объединяет строки входных таблиц. Давайте проверим форму исходных и составных таблиц, чтобы проверить операцию:

In [10]:
print('Shape of the `air_quality_pm25` table: ', air_quality_pm25.shape)

Shape of the `air_quality_pm25` table:  (1110, 4)


In [11]:
print('Shape of the `air_quality_no2` table: ', air_quality_no2.shape)

Shape of the `air_quality_no2` table:  (2068, 4)


In [12]:
print('Shape of the resulting `air_quality` table: ', air_quality.shape)

Shape of the resulting `air_quality` table:  (3178, 4)


Следовательно, результирующая таблица имеет `3178 = 1110 + 2068` строк.

Аргумент `axis` встречается в ряде методов, которые могут применяться вдоль оси. `DataFrame` имеет две соответствующие оси: первая, проходящая вертикально вниз по строкам (`ось 0`), и вторая, проходящая горизонтально по столбцам (`ось 1`). Большинство операций, таких как конкатенация или сводная статистика, по умолчанию выполняются по строкам (`ось 0`), но также могут применяться к столбцам.

Сортировка таблицы по дате и времени иллюстрирует также комбинацию обеих таблиц, причем столбец parameter определяет источник таблицы (либо `no2` из таблицы `air_quality_no2`, либо `pm25` из таблицы `air_quality_pm25`):

In [13]:
air_quality = air_quality.sort_values("date.utc")

In [14]:
air_quality.head()

date.utc            location parameter  value
2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0
1003  2019-05-07 01:00:00+00:00             FR04014       no2   25.0
100   2019-05-07 01:00:00+00:00             BETR801      pm25   12.5
1098  2019-05-07 01:00:00+00:00             BETR801       no2   50.5
1109  2019-05-07 01:00:00+00:00  London Westminster      pm25    8.0

В этом примере столбец `parameter`, позволяет идентифицировать каждую из исходных таблиц. Это не всегда так, функция `concat` предоставляет удобное решение с аргументом `keys`, добавляя дополнительный (иерархический) индекс строки. Например:

In [15]:
air_quality_ = pd.concat([air_quality_pm25, air_quality_no2],
                         keys=["PM25", "NO2"])

In [16]:
air_quality_.head()

date.utc location parameter  value
PM25 0  2019-06-18 06:00:00+00:00  BETR801      pm25   18.0
     1  2019-06-17 08:00:00+00:00  BETR801      pm25    6.5
     2  2019-06-17 07:00:00+00:00  BETR801      pm25   18.5
     3  2019-06-17 06:00:00+00:00  BETR801      pm25   16.0
     4  2019-06-17 05:00:00+00:00  BETR801      pm25    7.5

Существование нескольких индексов строк/столбцов одновременно не упоминалось ранее. Иерархическая индексация или `MultiIndex` - это продвинутая и мощная функция `pandas` для анализа многомерных данных.

На данный момент помните, что функцию `reset_index` можно использовать для преобразования любого уровня индекса в столбец, например, 

```Python
air_quality.reset_index(level=0)
```

Не стесняйтесь погрузиться в мир мультииндексирования в [разделе руководства пользователя по расширенной индексации](https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html#advanced).

Дополнительные параметры конкатенации таблиц (с точки зрения строк и столбцов) и того, как `concat` можно использовать для определения логики (объединения или пересечения) индексов на других осях, представлены в [разделе о конкатенации объектов](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html#merging-concat).

### Объединяйте таблицы, используя общий идентификатор

<img src="https://pandas.pydata.org/pandas-docs/stable/_images/08_merge_left.svg">

Координаты станции измерения качества воздуха хранятся в файле данных `air_quality_stations.csv`.

In [17]:
stations_coord = pd.read_csv("data/air_quality_stations.csv")
stations_coord.head()

location  coordinates.latitude  coordinates.longitude
0  BELAL01              51.23619                4.38522
1  BELHB23              51.17030                4.34100
2  BELLD01              51.10998                5.00486
3  BELLD02              51.12038                5.02155
4  BELR833              51.32766                4.36226

Станции, используемые в этом примере (`FR04014`, `BETR801` и `London Westminster`) - это всего лишь три записи в таблице метаданных. Мы хотим добавить координаты этих станций в таблицу измерений, каждая из которых находится в соответствующих строках таблицы `air_quality`.

In [18]:
air_quality.head()

date.utc            location parameter  value
2067  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0
1003  2019-05-07 01:00:00+00:00             FR04014       no2   25.0
100   2019-05-07 01:00:00+00:00             BETR801      pm25   12.5
1098  2019-05-07 01:00:00+00:00             BETR801       no2   50.5
1109  2019-05-07 01:00:00+00:00  London Westminster      pm25    8.0

Добавим координаты станции, предоставленные в таблице метаданных станций, в соответствующие строки таблицы измерений:

In [19]:
air_quality = pd.merge(air_quality, stations_coord,
                        how='left', on='location')
air_quality.head()

date.utc            location parameter  value  \
0  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0   
1  2019-05-07 01:00:00+00:00             FR04014       no2   25.0   
2  2019-05-07 01:00:00+00:00             FR04014       no2   25.0   
3  2019-05-07 01:00:00+00:00             BETR801      pm25   12.5   
4  2019-05-07 01:00:00+00:00             BETR801       no2   50.5   

   coordinates.latitude  coordinates.longitude  
0              51.49467               -0.13193  
1              48.83724                2.39390  
2              48.83722                2.39390  
3              51.20966                4.43182  
4              51.20966                4.43182

Используя функцию [`merge()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.merge.html#pandas.merge), для каждой строки таблицы `air_quality` добавляются соответствующие координаты из таблицы `air_quality_stations_coord`. Обе таблицы имеют общий столбец `location`, который используется в качестве ключа для объединения информации. Выбрав объединение `left`, в результирующей таблице `air_quality` окажутся только местоположения, доступные в (левой) таблице, например `FR04014`, `BETR801` и `London Westminster`. В функции merge поддерживает несколько опции, подобных операциям из базы данных.

Добавим описание и имя параметра, предоставленные таблицей метаданных, в таблицу измерений:

Метаданные параметров о качестве воздуха хранятся в файле `air_quality_parameters.csv`.

In [20]:
air_quality_parameters = pd.read_csv("data/air_quality_parameters.csv")
air_quality_parameters.head()

id                                        description  name
0    bc                                       Black Carbon    BC
1    co                                    Carbon Monoxide    CO
2   no2                                   Nitrogen Dioxide   NO2
3    o3                                              Ozone    O3
4  pm10  Particulate matter less than 10 micrometers in...  PM10

In [21]:
air_quality = pd.merge(air_quality, air_quality_parameters,
                       how='left', left_on='parameter', right_on='id')
air_quality.head()

date.utc            location parameter  value  \
0  2019-05-07 01:00:00+00:00  London Westminster       no2   23.0   
1  2019-05-07 01:00:00+00:00             FR04014       no2   25.0   
2  2019-05-07 01:00:00+00:00             FR04014       no2   25.0   
3  2019-05-07 01:00:00+00:00             BETR801      pm25   12.5   
4  2019-05-07 01:00:00+00:00             BETR801       no2   50.5   

   coordinates.latitude  coordinates.longitude    id  \
0              51.49467               -0.13193   no2   
1              48.83724                2.39390   no2   
2              48.83722                2.39390   no2   
3              51.20966                4.43182  pm25   
4              51.20966                4.43182   no2   

                                         description   name  
0                                   Nitrogen Dioxide    NO2  
1                                   Nitrogen Dioxide    NO2  
2                                   Nitrogen Dioxide    NO2  
3  Particulate matter less than 2.5 micrometers i...  PM2.5  
4                                   Nitrogen Dioxide    NO2

По сравнению с предыдущим примером нет общего имени столбца. Однако столбец parameter в таблице `air_quality` и `столбец id` в `air_quality_parameters` содержат переменную в общем формате. Аргументы `left_on` и `right_on` используются, чтобы сделать связь между двумя таблицами.

pandas поддерживают внутренние, внешние и правые соединения. Более подробная информация о `join/merge` таблиц представлена в [разделе руководства пользователя по объединению таблиц в стиле базы данных](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html#merging-join). Или взгляните на [страницу сравнения с SQL](https://pandas.pydata.org/pandas-docs/stable/getting_started/comparison/comparison_with_sql.html#compare-with-sql-join).

См. Руководство пользователя для [полного описания различных средств для объединения таблиц данных](https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html#merging).